src/basic/macro.h

```diff
@@ -463,4 +463,20 @@ typedef struct {
 
 assert_cc(sizeof(dummy_t) == 0);
 
+/* A little helper for subtracting 1 off a pointer in a safe UB-free way. This is intended to be used for for
+ * loops that count down from a high pointer until some base. A naive loop would implement this like this:
+ *
+ * for (p = end-1; p >= base; p--) …
+ *
+ * But this is not safe because p before the base is UB in C. With this macro the loop becomes this instead:
+ *
+ * for (p = PTR_SUB1(end, base); p; p = PTR_SUB1(p, base)) …
+ *
+ * And is free from UB! */
+#define PTR_SUB1(p, base)                                \
+        ({                                               \
+                typeof(p) _q = (p);                      \
```

Comment:
is `_q` here safe for nesting? Should it use `UNIT_T`?

Reply:
no, it's not safe for nesting. But I am not sure we should complicate stuff with the UNIQ_T stuff unless it's very likely or actually used in a nested way. At least I have serious trouble still following the macros if UNIQ_T is in the mix.

Hence, yes, UNIQ_T where necessary. But where it likely isn't, let's maybe avoid this until it becomes an issue or likely becomes an issue.

I mean, for loops are likely going to be nested, hence using UNIQ_T for the iterator variables of macros like STRV_FOREACH makes a ton of sense. But I doubt people will use `PTR_SUB1(PTR_SUB1(x, foo), bar)`. I mean, we have a pretty good idea here how this is going to be used, I even explained that in the comment, and there is no nesting there.

Thankfully all current compilers will generate very explicit warnings once you actually end up nesting this, so there aren't any mistakes possible here: if people actually nest it, the compiler will give you a clear hint that the macro first needs some love...

Reply:
makes sense